library/core/src/intrinsics/simd.rs

```diff
@@ -475,12 +475,19 @@ extern "rust-intrinsic" {
     /// * On little endian, the least significant bit corresponds to the first vector element.
     /// * On big endian, the least significant bit corresponds to the last vector element.
     ///
-    /// For example, `[!0, 0, !0, !0]` packs to `0b1101` on little endian and `0b1011` on big
-    /// endian.
-    ///
-    /// To consider a larger example, `[!0, 0, 0, 0, 0, 0, 0, 0, !0, !0, 0, 0, 0, 0, !0, 0]` packs
-    /// to `[0b00000001, 0b01000011]` or `0b0100001100000001` on little endian, and `[0b10000000,
-    /// 0b11000010]` or `0b1000000011000010` on big endian.
+    /// For example, `[!0, 0, !0, !0]` packs to
+    /// - `0b1101u8` or `[0b1101]` on little endian, and
+    /// - `0b1011u8` or `[0b1011]` on big endian.
+    ///
+    /// To consider a larger example,
+    /// `[!0, 0, 0, 0, 0, 0, 0, 0, !0, !0, 0, 0, 0, 0, !0, 0]` packs to
+    /// - `0b0100001100000001u16` or `[0b00000001, 0b01000011]` on little endian, and
+    /// - `0b1000000011000010u16` or `[0b10000000, 0b11000010]` on big endian.
+    ///
+    /// And finally, a non-power-of-2 example with multiple bytes:
+    /// `[!0, !0, 0, !0, 0, 0, !0, 0, !0, 0]` packs to
+    /// - `0b0101001011u16` or `[0b01001011, 0b01]` on little endian, and
+    /// - `0b1101001010u16` or `[0b11, 0b01001010]` on big endian.
```

Comment:
I don't have big-endian hardware to verify this, so this is based on Miri. However, AFAIK the intent is that the two ways of representing the result are equal-up-to-transmute, and it is clearly the case that `0b1101001010u16` transmutes to `[0b11, 0b01001010]`.

Reply:
Hmm, with big endian we've been having some interesting codegen bugs that we're looking into. I am hoping these don't put the lie to this remark, but...

Reply:
> I don't have big-endian hardware to verify this, so I could only validate this in Miri. However, AFAIK the intent is that the two ways of representing the result are equal-up-to-transmute, and it is clearly the case that `0b1101001010u16` transmutes to `[0b11, 0b01001010]`.

afaik the intent is actually to match what llvm does when bitcasting from e.g. `<10 x i1>` to `i10`.

you can likely use llvm const propagation to test it by compiling for some big-endian target and comparing the optimized llvm-ir to just returning a constant. alternatively, setting up a chroot with a different arch using debootstrap and qemu-user isn't that hard on linux, though llvm's backends may have bugs around weird vector lengths.

Reply:
Looking at [codegen for this intrinsic](https://github.com/rust-lang/rust/blob/a9702a66684f0e1f412d92a83761be2e461b69fb/compiler/rustc_codegen_llvm/src/intrinsic.rs#L1379-L1390), I am a bit confused: for an input vector of length 10, it seems to expect an `i10` return type -- but that's impossible (and contradicts the docs), the return type is `i16`. So I don't understand how [this case](https://github.com/rust-lang/rust/blob/a9702a66684f0e1f412d92a83761be2e461b69fb/compiler/rustc_codegen_llvm/src/intrinsic.rs#L1423-L1426) can ever be hit for larger-than-8 non-power-of-2 vectors.

However, for arrays things seem quite clear -- we first zero-extend to `expected_bytes * 8`, which in this case is i16 and will result in 0b0000001101001010, and then that gets transmuted to `[u8; 2]` (via alloca, store, load) which will result in `[0b00000011, 0b01001010]`.

Reply:
Something else is also odd about the implementation -- it goes through some lengths to specifically return the most significant bit (MSB) of each lane, but the docs for the intrinsic say that the input vector must contain only `0` and `!0` so we could just return any bit -- the lshr that the intrinsic generates seems entirely unnecessary.

Reply:
> So I don't understand how [this case](https://github.com/rust-lang/rust/blob/a9702a66684f0e1f412d92a83761be2e461b69fb/compiler/rustc_codegen_llvm/src/intrinsic.rs#L1423-L1426) can ever be hit for larger-than-8 non-power-of-2 vectors.

Lol, turns out it indeed cannot be hit. Miri implements the docs as written but codegen is buggy -- see [this test](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=af9900c577ccb5ccae6b5aef2841c6a5).

Reply:
Interestingly, simd_select_bitmask [works a bit differently](https://github.com/rust-lang/rust/blob/a9702a66684f0e1f412d92a83761be2e461b69fb/compiler/rustc_codegen_llvm/src/intrinsic.rs#L1122) and does accept `u16` here. So this is probably a bug in simd_bitmask?

Reply:
> Something else is also odd about the implementation -- it goes through some lengths to specifically return the most significant bit (MSB) of each lane, but the docs for the intrinsic say that the input vector must contain only `0` and `!0` so we could just return any bit -- the lshr that the intrinsic generates seems entirely unnecessary.

iirc that's because LLVM doesn't actually always know the elements are only `0` and `!0` (which imo llvm needs to gain new annotations for), so using the MSB bit allows it to optimize to `pmovmskb` and similar instructions.

Reply:
Oh I see, so adding more instructions makes it optimize better... but I assume for other targets that don't have a native "get MSB" instruction this could be a pessimization?

Reply:
well, [wasm has that operation](https://webassembly.github.io/spec/core/exec/instructions.html#t-mathsf-x-n-mathsf-xref-syntax-instructions-syntax-instr-vec-mathsf-bitmask) too, so it can't be too bad otherwise wasm wouldn't have added it.

Reply:
on aarch64, x86_64, and risc-v v it's better to have the shift that takes the msb bit: https://llvm.godbolt.org/z/rsrqrncvf